### DISCUSSION TITLE
Enhance Get-ChildItem so that it can filter recused directories names.

### DISCUSSION THREAD

**Discussion by MrDRamos** [PowerShell/PowerShell#15831] [Enhancements]
Enhance Get-ChildItem so that it can filter recused directories names.
Add 2 switches: -RecuseInclude & -RecurseExclude that operate on the recused directory names similar to how the -Include & -Exclude currently operate on the recused filenames.

Example: I want to find *.cs files in all subdirectories named `src*` but exclude any subdirectories names ending with `*test` or `*unit`
`Get-ChildItem -Recurse -RecurseInclude "src*" -RecurseExclude "*test","*unit" -Path ".\*" -Include "*.cs" -Exclude "test*" `

**Comment by iSazonov** [comment]
Main confusion here is what is "directory names" - short name or path? 
Really there are already some ways to do this by means of `Path` parameter and Where-Object cmdlet in pipeline. It is simple and clear. We could do something new only if we want more performance but I am not sure we can achieve this due to specific underlying .Net API.

**Reply by MrDRamos** [reply]
I meant the short folder name. The issue with piping is that you end up writing more error-prone code and you also have to fiddle around with regular expressions to the extract the folder name ... I don't think that performance is an issue, the typical use case would not be for 1000's of folders. But of way more importance is a clear expression of intent and a standard way of doing a common task. I don't want to search the web to find 10 different ways on how to write a compound pipped expression using  regex expressions I can't remember.

**Reply by iSazonov** [reply]
> I meant the short folder name.

From your proposal it is not clear. It is common problem with switch names. We have some issues about Include/Exclude parameter behavior.
To get reliable result we should point (1) filter pattern, (2) a kind of target (file or directory).

> I don't think that performance is an issue

Really it is an issue. FileSystemProvider has poor performance in many scenarios and we should take this into account.

> I don't want to search the web to find 10 different ways on how to write a compound pipped expression using regex expressions I can't remember.

In any case (in parameter or in pipeline) you have to use a filter (simple or Regex).

**Comment by MrDRamos** [comment]
I could think of introducing a new function whose sole purpose it is to return a list of paths via recursion e.g. Get-ParentItem. That path-list can then be passed to Get-ChildItem to retrieve the files. But Get-ChildItem could be used for both e.g:
```
$myPaths = Get-ChildItem -Directory -Recurse -RecurseInclude <list> -RecurseExclude <list> ...  
$myFiles = $myPaths | Get-ChildItem <file retrieval args>
```
Note that -Include/Exclude arguments can't be used to filter out directories. Therefore the need for new -RecurseInclude/Exclude arguments

**Reply by iSazonov** [reply]
For reference https://github.com/PowerShell/PowerShell/issues/15159
